Error processing tar file(archive/tar: missed writing 4096 bytes):

### Description
The `jkube:build` goal failes on `CentOS Linux release 7.9.2009 (Core)` with:

```
[ERROR] Failed to execute goal org.eclipse.jkube:kubernetes-maven-plugin:1.3.0:build (default) on project myapp-docker: Failed to execute the build: Error while trying to build the image: Unable to build image [someco/myapp:latest] : {"message":"Error processing tar file(archive/tar: missed writing 4096 bytes): "} (Internal Server Error: 500) -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.eclipse.jkube:kubernetes-maven-plugin:1.3.0:build (default) on project myapp-docker: Failed to execute the build
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:213)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:154)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:146)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:117)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:81)
        at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
        at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:309)
        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:194)
        at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:107)
        at org.apache.maven.cli.MavenCli.execute(MavenCli.java:993)
        at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:345)
        at org.apache.maven.cli.MavenCli.main(MavenCli.java:191)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
        at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
        at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
Caused by: org.apache.maven.plugin.MojoExecutionException: Failed to execute the build
        at org.eclipse.jkube.maven.plugin.mojo.build.AbstractDockerMojo.buildAndTag(AbstractDockerMojo.java:665)
        at org.eclipse.jkube.maven.plugin.mojo.build.AbstractDockerMojo.processImageConfig(AbstractDockerMojo.java:503)
        at org.eclipse.jkube.maven.plugin.mojo.build.AbstractDockerMojo.executeBuildGoal(AbstractDockerMojo.java:632)
        at org.eclipse.jkube.maven.plugin.mojo.build.BuildMojo.executeInternal(BuildMojo.java:49)
        at org.eclipse.jkube.maven.plugin.mojo.build.AbstractDockerMojo.execute(AbstractDockerMojo.java:444)
        at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
        ... 20 more
Caused by: org.eclipse.jkube.kit.config.service.JKubeServiceException: Error while trying to build the image: Unable to build image [someco/myapp:latest] : {"message":"Error processing tar file(archive/tar: missed writing 4096 bytes): "} (Internal Server Error: 500)
        at org.eclipse.jkube.kit.config.service.kubernetes.DockerBuildService.build(DockerBuildService.java:52)
        at org.eclipse.jkube.maven.plugin.mojo.build.AbstractDockerMojo.buildAndTag(AbstractDockerMojo.java:662)
        ... 26 more
Caused by: org.eclipse.jkube.kit.build.service.docker.access.DockerAccessException: Unable to build image [someco/myapp:latest] : {"message":"Error processing tar file(archive/tar: missed writing 4096 bytes): "} (Internal Server Error: 500)
        at org.eclipse.jkube.kit.build.service.docker.access.hc.DockerAccessWithHcClient.buildImage(DockerAccessWithHcClient.java:272)
        at org.eclipse.jkube.kit.build.service.docker.BuildService.doBuildImage(BuildService.java:179)
        at org.eclipse.jkube.kit.build.service.docker.BuildService.buildImage(BuildService.java:143)
        at org.eclipse.jkube.kit.build.service.docker.BuildService.buildImage(BuildService.java:77)
        at org.eclipse.jkube.kit.config.service.kubernetes.DockerBuildService.build(DockerBuildService.java:44)
        ... 27 more
Caused by: org.eclipse.jkube.kit.build.service.docker.access.hc.http.HttpRequestException: {"message":"Error processing tar file(archive/tar: missed writing 4096 bytes): "} (Internal Server Error: 500)
        at org.eclipse.jkube.kit.build.service.docker.access.hc.ApacheHttpClientDelegate$StatusCodeCheckerResponseHandler.handleResponse(ApacheHttpClientDelegate.java:200)
        at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:223)
        at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:165)
        at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:140)
        at org.eclipse.jkube.kit.build.service.docker.access.hc.ApacheHttpClientDelegate.post(ApacheHttpClientDelegate.java:121)
        at org.eclipse.jkube.kit.build.service.docker.access.hc.DockerAccessWithHcClient.buildImage(DockerAccessWithHcClient.java:270)
        ... 31 more

```

The same project is working on Windows.

The TAR file looks fine with `tar -tf docker-build.tar`.

The image can be built manually by going into the assembly directory and running `docker build .` so `docker` and the `docker daemon` are working.

The TAR file is somehow not valid for the `docker daemon`:

```
[root@srv3 tmp]# cat docker-build.tar | docker build -
Sending build context to Docker daemon   212 kB
Error response from daemon: Error processing tar file(archive/tar: missed writin 4096 bytes):
[root@srv3 tmp]#
```

### Info

* Eclipse JKube version : 1.3.0
* Maven version (`mvn -v`) : 3.6.3

@davidecavestro : I'm trying to reproduce your issue. I cloned https://github.com/davidecavestro/vagrant-jkube-connection-reset-by-peer adding fabric8-maven-plugin with the same configuration. But `mvn fabric8:build` is failing with this error:
```
[INFO] --- fabric8-maven-plugin:4.4.1:build (default-cli) @ test ---
[INFO] F8: Running in Kubernetes mode
[INFO] F8: Building Container image with Docker in Kubernetes mode
[INFO] Building tar: /home/rokumar/work/repos/vagrant-jkube-connection-reset-by-peer/target/docker/test/1.3.0/tmp/docker-build.tar
[INFO] F8: [test:1.3.0]: Created docker-build.tar in 21 milliseconds
[ERROR] F8: Failed to execute the build: io.fabric8.maven.docker.access.DockerAccessException: Unable to build image [test:1.3.0] : "COPY failed: file not found in build context or excluded by .dockerignore: stat maven: file does not exist" 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  2.071 s
[INFO] Finished at: 2021-07-23T15:17:22+05:30
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal io.fabric8:fabric8-maven-plugin:4.4.1:build (default-cli) on project test: Failed to execute the build: io.fabric8.maven.docker.access.DockerAccessException: Unable to build image [test:1.3.0] : "COPY failed: file not found in build context or excluded by .dockerignore: stat maven: file does not exist"  -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException

```

So I've tried the vagrant box sample with 1.4.0, but as soon as I add the deploy step just to simulate real-world scenario i.e. 
`mvn -DskipTests clean package k8s:build deploy -DaltDeploymentRepository=snapshot-repo::default::file:/tmp/foo`
I got 

> [ERROR] Caused by: java.lang.IllegalArgumentException: Can not set org.eclipse.aether.spi.log.Logger field org.apache.maven.repository.internal.DefaultVersionResolver.logger to org.eclipse.aether.internal.impl.slf4j.Slf4jLoggerFactory
[ERROR] at sun.reflect.UnsafeFieldAccessorImpl.throwSetIllegalArgumentException(UnsafeFieldAccessorImpl.java:167)
[ERROR] at sun.reflect.UnsafeFieldAccessorImpl.throwSetIllegalArgumentException(UnsafeFieldAccessorImpl.java:171)
[ERROR] at sun.reflect.UnsafeObjectFieldAccessorImpl.set(UnsafeObjectFieldAccessorImpl.java:81)
[ERROR] at java.lang.reflect.Field.set(Field.java:764)
[ERROR] at org.eclipse.sisu.bean.BeanPropertyField.set(BeanPropertyField.java:72)

@pmbsa : I'm trying to reproduce your issue in Vagrant box from https://github.com/davidecavestro/vagrant-jkube-connection-reset-by-peer.

When I run `mvn k8s:build` in your project I'm seeing this error:
```
[ERROR] k8s: Failed to execute the build [Error while trying to build the image: Unable to build image [jkube/context-and-assembly] : Connection reset by peer]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  02:04 min
[INFO] Finished at: 2022-06-24T13:44:27+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.eclipse.jkube:kubernetes-maven-plugin:1.5.1:build (default-cli) on project docker-file-provided: Failed to execute the build: Error while trying to build the image: Unable to build image [jkube/context-and-assembly] : Connection reset by peer -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

Were you seeing the same issue with your project?

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -21,6 +21,7 @@ Usage:
 ./scripts/extract-changelog-for-version.sh 1.3.37 5
 ```
 ### 1.9.0-SNAPSHOT
+* Fix #777: `k8s:build` with Dockerfile throws `Connection reset by peer` error on old docker daemons
 * Fix #1279: Remove redundant log messages regarding plugin modes
 * Fix #1438: Add configuration option in ServiceAccountEnricher to skip creating ServiceAccounts
 * Fix #1468: Add startup probe in QuarkusHealthCheckEnricher

--- a/jkube-kit/common/src/main/java/org/eclipse/jkube/kit/common/archive/JKubeTarArchiver.java
+++ b/jkube-kit/common/src/main/java/org/eclipse/jkube/kit/common/archive/JKubeTarArchiver.java
@@ -80,8 +80,10 @@ public class JKubeTarArchiver {
         if (fileModeMap.containsKey(currentFile)) {
           tarEntry.setMode(Integer.parseInt(fileModeMap.get(currentFile), 8));
         } else if (currentFile.isDirectory()) {
-          tarEntry.setSize(0L);
           tarEntry.setMode(TarArchiveEntry.DEFAULT_DIR_MODE);
+        }
+        if (currentFile.isDirectory()) {
+          tarEntry.setSize(0L);
         }
         Optional.ofNullable(tarArchiveEntryCustomizer).ifPresent(tac -> tac.accept(tarEntry));
         tarArchiveOutputStream.putArchiveEntry(tarEntry);

--- a/jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/archive/JKubeTarArchiverTest.java
+++ b/jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/archive/JKubeTarArchiverTest.java
@@ -17,6 +17,8 @@ import java.io.File;
 import java.io.IOException;
 import java.nio.charset.StandardCharsets;
 import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
 import java.util.function.Consumer;
 
 import org.eclipse.jkube.kit.common.assertj.ArchiveAssertions;
@@ -193,6 +195,32 @@ public class JKubeTarArchiverTest {
         .containsExactlyInAnyOrder(
             tuple("file.txt", 12L, defaultFileMode),
             tuple("nested/directory/" + LONG_FILE_NAME, 19L, defaultFileMode)
+        );
+  }
+
+  @Test
+  public void createTarBall_defaultCompressionWithFileModes_createsTarWithCorrectEntrySizeAndModeForFiles() throws Exception {
+    // Given
+    final int defaultDirMode = Integer.parseInt("040755", 8); // 16877
+    final File outputFile = temporaryFolder.newFile("target.noExtension");
+    Map<File, String> fileModeMap = new HashMap<>();
+    fileModeMap.put(new File(toCompress, "nested"), "040755");
+    fileModeMap.put(new File(toCompress, "nested/directory"), "040755");
+
+    // When
+    final File result = JKubeTarArchiver.createTarBall(outputFile, toCompress, FileUtil.listFilesAndDirsRecursivelyInDirectory(toCompress), fileModeMap, ArchiveCompression.none);
+
+    // Then
+    ArchiveAssertions.assertThat(result)
+        .isSameAs(outputFile)
+        .isNotEmpty()
+        .entries()
+        .filteredOn(TarArchiveEntry::isDirectory)
+        .hasSize(2)
+        .extracting("name", "size", "mode")
+        .containsExactlyInAnyOrder(
+            tuple("nested/", 0L, defaultDirMode),
+            tuple("nested/directory/", 0L, defaultDirMode)
         );
   }
 }

